### PR TITLE
Integrations Identity pipeline refactoring

### PR DESCRIPTION
**Pull request by evanpurkhiser**
Probably easiest to review the 4 commits individually.

This is a large refactoring of the v3 Integration setup pipeline to use a new style of user identity lookup, based on a new `identity` module, which provides a pipeline + identity providers.

This is the foundation for unifying the SSO Authentication identity, integrations identity, and (at some point probably in the far future) plugin identity (likely when they're moved to new-style integrations).

This implements the following:

 - The `identity` module, which provides an Identity pipeline, with a generic identity association `finish_pipeline` implementation. This may be used both directly as part of an identity association endpoint, as well as made part of a pipeline through the `pipleline.NestedProviderView`. A generic base `OAuth2` provider is also implemented, which is mostly a copy and paste from both the `auth` module and `integration` modules implementations (which this will replace).

 - The OAuth identity provider supports scope customization through the pipeline config object.

 - The `models.identity` module has been updated with models that more accurately reflect how Identity is stored.

   - One identity per user

   - `IdentityProvider` is associated to an organization, allowing lookup of identity providers configured for an organization. Originally the idea was to have one identity provider per 'instance' of the identity provider, for example there'd be one Slack IDP since there's only one Sentry app on slack. However, with this implementation we'd stil need some way to know it's configured for an organization, and that linking table would likely take just as much space. Further, having a direct association to an identity provider makes generic identity association possible.

   - Also **Important to note** is that these tables were not yet in use outside of the feature-flagged v3 integrations, so there are two migrations here, the first simply removes the tables, the second re-creates the tables using the updated model fields.

 - The Integrations base object has it's `id` class attribute renamed to `key`. This is a little more clear as it differentiates it from the provider models database id.

 - Integrations now compose the slack identity provider pipeline to do setup and store the setup users identity.

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Changes require @getsentry/security sign-off</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR includes migrations</td>
    </tr>
  </tbody>
</table>

### Security concerns found

- src/sentry/{integrations/oauth.py => identity/oauth2.py}


## Migration Checklist

- [ ] new columns need to be nullable (unless table is new)
- [ ] migration with any new index needs to be done concurrently
- [ ] data migrations should not be done inside a transaction
- [ ] before merging, check to make sure there aren't conflicting migration ids


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by EvanPurkhiser** [comment]
Assigning @dcramer since I know he knows this part of the code, but I know he's busy, so also assigning @alex-hofsteede for that knowledge share.

File: src/sentry/identity/__init__.py
Comment (alex-hofsteede) [inline src/sentry/identity/__init__.py]: does this need to be addressed before merge?

File: src/sentry/integrations/slack/integration.py
Comment (alex-hofsteede) [inline src/sentry/integrations/slack/integration.py]: this was done right?

File: tests/sentry/web/frontend/test_integration_setup.py
Comment (alex-hofsteede) [inline tests/sentry/web/frontend/test_integration_setup.py]: Do we need any more tests aside from the ones refactored here?

File: src/sentry/identity/base.py
Comment (EvanPurkhiser) [reply src/sentry/identity/base.py]: Yeah, I actually introduced this over here https://github.com/getsentry/sentry/pull/6577, but am moving it into the identity module, and it will be removed from the auth provider later once the auth providers make use of the identity provider.

File: src/sentry/identity/manager.py
Comment (EvanPurkhiser) [reply src/sentry/identity/manager.py]: This manager was just copied over from other provider managers. The auth module and integration module both have one of these. At some point we'll likely want to just provide some general lookup manager that can be used to this fashion. It doesn't have anything specific to identity providers.

File: src/sentry/identity/__init__.py
Comment (EvanPurkhiser) [reply src/sentry/identity/__init__.py]: I've updated the comment for this, but basically the idea is that this registering of the provider should happen with the slack plugin most likely.

File: src/sentry/integrations/slack/integration.py
Comment (EvanPurkhiser) [reply src/sentry/integrations/slack/integration.py]: Yeah, updated. good catch.

File: tests/sentry/web/frontend/test_integration_setup.py
Comment (EvanPurkhiser) [reply tests/sentry/web/frontend/test_integration_setup.py]: I have some identity linking acceptance tests that I need to finish up, I'll try and get those in here, but I'd like to get this merged in sooner than later, I'll try and get those in, but otherwise they'll come in as a separate PR I think.

Other than that more tests will show up as we solidify more of what integration setup looks like from a configuration standpoint

**Review by alex-hofsteede — APPROVED** [review]
LGTM

File: src/sentry/identity/slack/provider.py
Comment (mattrobenolt) [inline src/sentry/identity/slack/provider.py]: Set `flags=FLAG_PRIORITIZE_DISK` at least on these since I assume we'll be hardcoding these for us in config.

File: src/sentry/integrations/pipeline.py
Comment (mattrobenolt) [inline src/sentry/integrations/pipeline.py]: Use `sentry.utils.json` instead unless there was a reason not to.

File: src/sentry/models/identity.py
Comment (mattrobenolt) [inline src/sentry/models/identity.py]: Let's not actually drop this yet until a later migration.

File: src/sentry/models/identity.py
Comment (mattrobenolt) [inline src/sentry/models/identity.py]: This needs to be nullable.

File: src/sentry/models/identity.py
Comment (mattrobenolt) [inline src/sentry/models/identity.py]: Ok, this migration is going to be problematic.

We need to come up with a transition plan. We can't drop and add these new columns at the same time, nor can we simply edit this unique_together, or add not nullable columns.

**Review by mattrobenolt — CHANGES_REQUESTED** [review]
This migration is a blocker. Let's chat IRL to come up with the plan.

**Comment by EvanPurkhiser** [comment]
Talked IRL with @mattrobenolt. I've updated the description of the PR to note that the Identity models where not yet used anywhere outside of the V3 integration work, so the changes that were made to these tables can be safely be applied by dropping and recreating these tables as is done n the migrations here.

I will still however need to update the IDs of the migrations.

File: src/sentry/models/identity.py
Comment (EvanPurkhiser) [reply src/sentry/models/identity.py]: [Addressed in this comment](https://github.com/getsentry/sentry/pull/6708#issuecomment-354855601)

File: src/sentry/models/identity.py
Comment (EvanPurkhiser) [reply src/sentry/models/identity.py]: [Addressed in this comment](https://github.com/getsentry/sentry/pull/6708#issuecomment-354855601)

File: src/sentry/models/identity.py
Comment (EvanPurkhiser) [reply src/sentry/models/identity.py]: [Addressed in this comment](https://github.com/getsentry/sentry/pull/6708#issuecomment-354855601)

File: src/sentry/identity/slack/provider.py
Comment (EvanPurkhiser) [reply src/sentry/identity/slack/provider.py]: Yeah these will be hardcoded in the configs. [Updated here](https://github.com/getsentry/sentry/pull/6708/files#diff-dc19567bb1802d08f2e32eacb8e0980dR7).

File: src/sentry/integrations/pipeline.py
Comment (EvanPurkhiser) [reply src/sentry/integrations/pipeline.py]: [Updated here](https://github.com/getsentry/sentry/pull/6708/files#diff-94419ddbfb31de69ee4e9fefa0cd2a87R11).

**Comment by EvanPurkhiser** [comment]
@mattrobenolt: bumped my migrations.

**Comment by EvanPurkhiser** [comment]
Had to bump the migrations one more time because of https://github.com/getsentry/sentry/pull/6147. Just need your sign off @mattrobenolt.

**Review by mattrobenolt — APPROVED** [review]
Punch it.

**Pull request merged**
